### DISCUSSION TITLE
Replacement for pydantic.json.pydantic_encoder

### DISCUSSION THREAD

**Discussion by anneum** [pydantic/pydantic#6652] [Question] [answered]
First of all, congratulations on the great job you did with pydantic v2.
With pydantic v1 I used the method ``pydantic.json.pydantic_encoder`` to serialize my sqlalchemy queries or as in the example here to send a request with the serialized data. Is it possible to do this with pydantic v2 and if so, how?

```python
import json
import aiohttp
import pydantic
from pydantic import BaseModel

def _custom_json_serializer(*args, **kwargs) -> str:  # type: ignore
    """Encodes json in the same way that pydantic does."""
    return json.dumps(*args, default=pydantic.json.pydantic_encoder, **kwargs)

async def post_datapoint(datapoint: BaseModel, endpoint: str) -> None:
    async with aiohttp.ClientSession(json_serialize=_custom_json_serializer) as session:
        async with session.post(endpoint, json=datapoint) as response:
            print(response)
```

**Comment by jimmyruann** [comment]
I ran into this issue, I wrote my own custom JSON encoder as a quick fix.  

```
class CustomPydanticJSONEncoder(json.JSONEncoder):
    def default(self, obj: Any) -> Any:
        if isinstance(obj, BaseModel):
            return obj.model_dump(mode="json")
        return super().default(obj)

models = [...]
json.dumps(models, cls=CustomPydanticJSONEncoder)
```

**Comment by lucafaggianelli** [comment]
If you're using fastapi you can use its `jsonable_encoder`

```py
from fastapi.encoders import jsonable_encoder
```

**Comment by jackklika** [comment]
This is especially important for the temporal python sdk for reliable encoding of temporal payloads. `fastapi.encoders.jsonable_encoder` should be a good replacement. But it would be helpful to add some guidance to the V1>V2 migration guide.

**Comment by tekumara** [comment, marked as answer]
See [pydantic_core.to_jsonable_python](https://docs.pydantic.dev/latest/api/pydantic_core/#pydantic_core.to_jsonable_python) to covert to a python objects (list, dict, etc) which can be fed to `json.dumps` to produce a json str or [pydantic_core.to_json](https://docs.pydantic.dev/latest/api/pydantic_core/#pydantic_core.to_json) to covert to json bytes.

**Comment by anneum** [comment]
Great, thank you very much.
For those who find the issue via the search function. Here is the code as I do the serialization with sqlalchemy.

```python
import json
from typing import Any

from pydantic_core import to_jsonable_python
from sqlalchemy.ext.asyncio import create_async_engine


def json_serializer(obj: Any) -> str:
    return json.dumps(to_jsonable_python(obj))


engine = create_async_engine("DATABASE_URL", json_serializer=json_serializer)
```

**Reply by tekumara** [reply]
This will double json encode I think - ie: you don't need the json.dumps

**Reply by anneum** [reply]
I think we need to do both here because according to the sqlalchemy [documentation](https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine.params.json_serializer) a "Python callable that will render a given object as JSON" is expected. As far as I understand, I get a python object from `pydantic_core.to_jsonable_python`. Or am I getting this wrong?

Addendum: Or should I use `pydantic_core.to_json` directly?

**Reply by tekumara** [reply]
You're correct - my apologies!

`pydantic_core. to_jsonable_python` produces python objects (eg: list, dict, etc.) from basemodels. To get a json str from that, using `json.dumps`. 

`pydantic_core.to_json` produces json bytes. To get a json str from that, using `decode`.

The only difference is the whitespace in the json str, ie: 
```
>>> pydantic_core.to_json(obj).decode("utf-8") == json.dumps(pydantic_core.to_jsonable_python(obj), separators=(',', ':'))
True
```

I've edited my answer above to correct this.

PS: I've not tested this with dates.

**Reply by martimors** [reply]
Nice example, but why did you have to set `pool_pre_ping` in this case?

**Reply by anneum** [reply]
@martimors Sorry, the `pool_pre_ping` is just a remnant of my code and has nothing to do with json encoding. I will edit my answer immediately.
